Add Oxford Flowers 102 Image Classification Dataset

DerrickUnleashed opened the pull request:
## This PR introduces support for the **Oxford Flowers 102** image classification dataset.

-  Implemented `flowers102_dataset()` loader with consistent design
-  Reuses the existing `download_and_cache()` utility for consistency and caching.
-  Adds a corresponding test suite, consistent with other image classification datasets.
-  Includes full Roxygen documentation with usage example and annotation support.

Closes #136

cregouby commented on R/dataset-flowers.R:
```diff
@@ -0,0 +1,256 @@
+#'
+#' @examples
+#' \dontrun{
+#' root_dir <- tempfile()
+#' flowers <- flowers102_dataset(root = root_dir, split = "train", download = TRUE)
```
**todo**: please use `root` for `root=`

DerrickUnleashed replied:
Thank you for the suggestion. I have updated the documentation to remove ambiguity.

cregouby commented on R/dataset-flowers.R:
```diff
@@ -0,0 +1,256 @@
+    "osteospermum",
+    "spring crocus",
+    "bearded iris",
+    "windflower",
+    "tree poppy",
+    "gazania",
+    "azalea",
+    "water lily",
+    "rose",
+    "thorn apple",
+    "morning glory",
+    "passion flower",
+    "lotus",
+    "toad lily",
+    "anthurium",
+    "frangipani",
+    "clematis",
+    "hibiscus",
+    "columbine",
+    "desert-rose",
+    "tree mallow",
+    "magnolia",
+    "cyclamen",
+    "watercress",
+    "canna lily",
+    "hippeastrum",
+    "bee balm",
+    "ball moss",
+    "foxglove",
+    "bougainvillea",
+    "camellia",
+    "mallow",
+    "mexican petunia",
+    "bromelia",
+    "blanket flower",
+    "trumpet creeper",
+    "blackberry lily"
+    )
+
+
```
**question** why not reusing the existing classes ?

DerrickUnleashed replied:
You're right—redefining the classes list inside `initialize()` is redundant since it's already declared at the top level of the `dataset()` constructor. Thanks for pointing this out.

cregouby commented on R/dataset-flowers.R:
```diff
@@ -0,0 +1,256 @@
+
+    img <- magick::image_read(img_path)
+    img <- magick::image_resize(img, paste(self$image_size, collapse = "x"))
+    img_tensor <- torchvision::transform_to_tensor(img)
```
**question**:  why wouldn't you keep it as array as for cifar and other datasets ?

DerrickUnleashed replied:
## Drawbacks of using the CIFAR-style (array-based) method for Flowers102:
- High Memory Usage: Unlike CIFAR's `32×32` images, Flowers102 images are high-resolution JPEGs. Even after resizing (e.g., to `224×224`), loading all images into memory as arrays can consume 1–2GB+ RAM per split.
- Longer Processing Time: The initial `download()` step would become significantly slower, as all images must be decoded, resized, and converted upfront.
- Reduced Flexibility: Preloading removes the ability to apply different transformations dynamically at runtime (e.g., augmentations).
- Less Efficient for Lazy Pipelines: The current path-based approach enables on-demand loading, which is more scalable for larger datasets or limited-resource systems.

That said, if you strongly prefer the CIFAR-style array-based design to maintain consistency, I can implement it accordingly, just let me know!

DerrickUnleashed replied:
- You're right — I was wrong earlier to return it in `tensor()` format directly. I've now changed the code so that it’s consistent with how `cifar10_dataset()` works.
- The image is kept as a plain array in [H, W, C] format, and it's only converted to a tensor if the user explicitly provides `transform = transform_to_tensor`.

cregouby commented on R/dataset-flowers.R:
```diff
@@ -0,0 +1,256 @@
+    rlang::inform("Extracting archive and processing metadata...")
+    untar(file.path(self$raw_folder, "102flowers.tgz"), exdir = self$raw_folder)
+
+    labels <- R.matlab::readMat(file.path(self$raw_folder, "imagelabels.mat"))$labels
+    setids <- R.matlab::readMat(file.path(self$raw_folder, "setid.mat"))
```
**todo** you can't use `R.matlab::` as it is not in the dependancies.
**suggestion** as it is very specific, you can use the requireNamespace trick like in
```
  if (!requireNamespace("R.matlab", quietly = TRUE))
    runtime_error("Package {.pkg R.matlab} is needed for this dataset. Please install it.")
}
```

DerrickUnleashed replied:
Thanks for the suggestion! I've added the `requireNamespace("R.matlab", quietly = TRUE)` along with a runtime error message if it's missing. This avoids adding it as a hard dependency while still handling the requirement gracefully.

DerrickUnleashed replied:
I have implemented the following
```r
    if (!requireNamespace("R.matlab", quietly = TRUE)) {
      runtime_error("Package 'R.matlab' is needed for this dataset. Please install it.")
    }
    labels <- R.matlab::readMat(file.path(self$raw_folder, "imagelabels.mat"))$labels
    setids <- R.matlab::readMat(file.path(self$raw_folder, "setid.mat"))
```
Yet the workflow runs into the same error.

I have now added `R.matlab` to the `DESCRIPTION` file to check if that runs without errors.

cregouby commented on tests/testthat/test-dataset-flowers.R:
**todo** Please test using the dataset in a dataloader.

DerrickUnleashed replied:
Thanks for the input. I have added a dataloader test.

cregouby reviewed (changes requested):
**todo** : Please check also the remarks made in #156 like
**improvement** : please refactor with a glue:: command for readability and consistency.

DerrickUnleashed commented:
@cregouby This PR is ready for review. I have made it such that it is consistent with the `cifar10_dataset()`. Do let me know if there's anything that needs to be changed.

cregouby commented:
> I have made changes in this PR in relation to the FER-2013 dataset loader PR to reduce the review efforts.
> 
>     * Refactor of rlang::inform to cli::cli_inform
> 
> I do however need to move from magick to jpeg.

Please use the internal `cli_inform` that is improved over `cli::cli_inform`. Same for other `cli_` messaging functions. (see `R/condition.R` )

DerrickUnleashed commented:
I have made the following changes:
- Converted to the use of internal `cli_` instead of `cli::cli_`
- Converted to the use of `cli_abort()` instead of `runtime_error()`
- The returned array for `$x` returns a array of `{C x H x W}` instead of `{H x W x C}` as suggested in `caltech101_dataset()`
- Moved from use of `magick::` to `jpeg::`.
- Split tests into batches as suggested in `caltech101_dataset()`
```r
flowers102 <- flowers102_dataset(download = TRUE)
first_item <- flowers102[100]
cat("Image Dims:",dim(first_item[[1]]),"\n") #3 500 666 
cat("Label:", first_item$y, "\n") #10 
cat("Class:",flowers102$classes[first_item$y],"\n") #globe thistle
```

DerrickUnleashed commented:
This is in relation to the Improvement suggested in `caltech101_dataset()`:

**Context :** `dataloader(caltech101_dataset(root = t, download = TRUE, transform = transform_to_tensor))`, `first batch$x` must be an image tensor of shape `c(1,3, x, y)` which is not the case...This issue will also arise in `flower102_dataset()`

**Explanation :** If the `first_item$x` returns an array of shape `{H x W x C}` then if I pass `transform = transform_to_tensor` it returns `first_item$x` as a tensor of shape `{C x H x W}`.
But as suggested earlier I have updated the `first_item$x` to return an array of shape `{C x H x W}` so now if i pass `transform = transform_to_tensor` it returns `first_item$x` as a tensor of shape `{W x C x H}`.

**Question :** Is it possible for us to return `first_item$x` as an array of shape `{H x W x C}` or do I need to write an if condition for when `transform = transform_to_tensor` is passed to change the shape ?

cregouby commented:
> This is in relation to the Improvement suggested in `caltech101_dataset()`:
> 
> **Context :** `dataloader(caltech101_dataset(root = t, download = TRUE, transform = transform_to_tensor))`, `first batch$x` must be an image tensor of shape `c(1,3, x, y)` which is not the case...This issue will also arise in `flower102_dataset()`
> 
> **Explanation :** If the `first_item$x` returns an array of shape `{H x W x C}` then if I pass `transform = transform_to_tensor` it returns `first_item$x` as a tensor of shape `{C x H x W}`. But as suggested earlier I have updated the `first_item$x` to return an array of shape `{C x H x W}` so now if i pass `transform = transform_to_tensor` it returns `first_item$x` as a tensor of shape `{W x C x H}`.
> 
> **Question :** Is it possible for us to return `first_item$x` as an array of shape `{H x W x C}` or do I need to write an if condition for when `transform = transform_to_tensor` is passed to change the shape ?

You are right, the dataset item$x is usually an array **as it comes out of the readJPEG()**, where aplying the recommanded (but not default) `transform = transform_to_tensor` **must** return a channel-first image-tensor. Sorry for my misleading comment (from a personnal confusion between the two)

DerrickUnleashed commented on R/dataset-flowers.R:
```diff
@@ -0,0 +1,175 @@
+    label_idx <- self$labels[[index]]
+    label <- self$classes[label_idx]
+
+    img <- jpeg::readJPEG(img_path)
```
Sure. I'll update `x` and `y` as soon as possible.

cregouby commented on R/dataset-flowers.R:
```diff
@@ -0,0 +1,175 @@
+#' Oxford Flowers 102 Dataset
+#'
+#' Loads the Oxford 102 Category Flower Dataset. This dataset consists of 102 flower categories, with between 40 and 258 images per class. 
```
**improvement** **blocking** we should have an explicit mention of the **variable size** of the images
**improvement** we should mension the task for the dataset : this is a **classification** dataset

DerrickUnleashed replied:
Thanks for the improvement. I agree we need to explicitly mention about the variable size of images and the task of the dataset. I'll update the documentation accordingly.

cregouby commented on R/dataset-flowers.R:
```diff
@@ -0,0 +1,175 @@
+#'   xs <- torch_stack(xs)
+#'   ys <- sapply(batch, function(sample) sample$y)
+#'   list(x = xs, y = ys)
+#' }
```
**improvement** remove this very complex part of the example 
**suggestion** replace by a more straithgful transform in the dataloader (see above)

DerrickUnleashed replied:
Thanks for the suggestion. I have now updated it to inline transforms and removed the need for separate function as per your suggestion.

cregouby commented on R/dataset-flowers.R:
```diff
@@ -0,0 +1,175 @@
+#'
+#' @examples
+#' \dontrun{
+#' flowers <- flowers102_dataset(split = "train", download = TRUE)
```
```suggestion
#' flowers <- flowers102_dataset(split = "train", download = TRUE,
#'     transform = . %>% transform_to_tensor %>% transform_resize(c(224, 224))
#'     )
```

DerrickUnleashed replied:
Thanks for the suggestion. I really like the inline transforms. I'll make sure to use these instead of a separate function. Hope the following example is as per expectations.
```r
# Load the dataset with inline transforms
flowers <- flowers102_dataset(
  split = "train",
  download = TRUE,
  transform = function(x) {
    x %>%
      transform_to_tensor() %>%
      transform_resize(c(224, 224))
  }
)

# Create a dataloader with default collate_fn
dl <- dataloader(
  dataset = flowers,
  batch_size = 4
)

# Access a batch
batch <- dataloader_next(dataloader_make_iter(dl))
batch$x  # Should be of shape c(4, 3, 224, 224)
batch$y  # Should be a list of 4 integer labels
```

DerrickUnleashed commented on R/dataset-flowers.R:
```diff
@@ -0,0 +1,175 @@
+#' - `"test"`: test subset with labels (used for evaluation).
+#'
+#' @inheritParams mnist_dataset
+#' @param root (string, optional): Root directory for dataset storage,
+#' the dataset will be stored under `root/flowers102`.
+#' @param split (string, optional): One of `"train"`, `"val"`, or `"test"`,
+#' defines which subset of the data to load. Default is `"train"`.
```
Thanks for the improvement. I'll make it less pythonic. Thing is I inherited the params from `mnist_dataset()` which had the following format that's why I tried to make it consistent. But now I have inherited from `fgvc_aircraft_dataset()` which is has what you expect.
```r
mnist_dataset()
#' @param root (string): Root directory of dataset where
#'   `MNIST/processed/training.pt` and  `MNIST/processed/test.pt` exist.
#' @param train (bool, optional): If True, creates dataset from
#'   `training.pt`, otherwise from `test.pt`.
#' @param download (bool, optional): If true, downloads the dataset from the
#'   internet and puts it in root directory. If dataset is already downloaded,
#'   it is not downloaded again.
#' @param transform (callable, optional): A function/transform that  takes in an
#'   PIL image and returns a transformed version. E.g,
#'   [transform_random_crop()].
#' @param target_transform (callable, optional): A function/transform that takes
#'   in the target and transforms it.
fgvc_aircraft_dataset()
#' @param root Character. Root directory for dataset storage. The dataset will be stored under `root/fgvc-aircraft-2013b`.
#' @param split Character. One of `"train"`, `"val"`, `"trainval"`, or `"test"`. Default is `"train"`.
#' @param annotation_level Character. Level of annotation to use for classification. Default is `"variant"`.
#' One of `"variant"`, `"family"`, `"manufacturer"`, or `"all"`. See *Details*.
#' @param transform Optional function to transform input images after loading. Default is `NULL`.
#' @param target_transform Optional function to transform labels. Default is `NULL`.
#' @param download Logical. Whether to download the dataset if not found locally. Default is `FALSE`.
```
